### PR TITLE
Add additional properties callback in Inspector Serialization Delegate

### PR DESCRIPTION
**Pull request by albertusdev**
## Description

Added callback to add additional experimental serialization properties.
This callback can be used to customize the serialization of DiagnosticsNode objects for experimental features. For example, Dart DevTools can evaluate the following expression to register a VM Service API with a custom serialization to experiment with visualizing layouts.

## Related Issues

-

## Tests
I added the following tests:
- `InspectorSerializationDelegate addAdditionalPropertiesCallback`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 1]
```diff
@@ -2914,6 +2915,8 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
   @override
   final bool expandPropertyValues;
 
+  final Map<String, Object> Function(DiagnosticsNode, _SerializationDelegate) addAdditionalPropertiesCallback;
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: Suggested documentation:
```
/// Callback to add additional experimental serialization properties.
///
/// This callback can be used to customize the serialization of DiagnosticsNode
/// objects for experimental features. For example, Dart DevTools can evaluate
/// the following expression to register a VM Service API with a custom serialization
/// to experiment with visualizing layouts.
/// *insert your example code here*
```

**Review by jacob314 — COMMENTED** [review]
Add a test to widget_inspector_test.dart
verifying the callback works as expected. You will find some existing code that tests the service extensions. You could write a test that simulates registering a custom service extension like the one you need to register from DevTools. In widget_inspector_test.dart, the service extension logic is stubbed out so it is easy to trigger service extensions and validate the responses.

**Comment by albertusangga** [comment]
@jacob314 ready for your another round of review before requesting review from Flutter team, might need your suggestion on the docs as well since we make the SerializationDelegate public now

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 2]
```diff
@@ -2914,8 +2927,50 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
...
+  /// Callback to add additional experimental serialization properties.
+  ///
+  /// This callback can be used to customize the serialization of DiagnosticsNode
+  /// objects for experimental features. For example, Dart DevTools can evaluate
+  /// the following expression to register a VM Service API with a custom serialization
+  /// to experiment with visualizing layouts.
+  ///
+  /// Example usage:
+  ///
+  /// Map<String, Object> getDetailsSubtreeWithRenderObject(
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: wrap this code example with 
"```dart"

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 3]
```diff
@@ -2914,8 +2927,50 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
...
   final bool expandPropertyValues;
 
+  /// Callback to add additional experimental serialization properties.
+  ///
+  /// This callback can be used to customize the serialization of DiagnosticsNode
+  /// objects for experimental features. For example, Dart DevTools can evaluate
+  /// the following expression to register a VM Service API with a custom serialization
+  /// to experiment with visualizing layouts.
+  ///
+  /// Example usage:
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: Replace "Example usage:" with
The following code samples demonstrates adding the RenderObject associated with an Element to the
serialized data for all elements in the tree.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 4]
```diff
@@ -2914,8 +2927,50 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
   @override
   final bool expandPropertyValues;
 
+  /// Callback to add additional experimental serialization properties.
+  ///
+  /// This callback can be used to customize the serialization of DiagnosticsNode
+  /// objects for experimental features. For example, Dart DevTools can evaluate
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: replace 
"Dart Devtools"
with
`[Dart Devtools](https://github.com/flutter/devtools)`

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 5]
```diff
@@ -2889,20 +2889,33 @@ int _toLocationId(_Location location) {
   return id;
 }
 
-class _SerializationDelegate implements DiagnosticsSerializationDelegate {
-  _SerializationDelegate({
+/// A delegate that configures how a hierarchy of [DiagnosticsNode]s should be
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: Add a `@visibleForTesting` annotation.

Tweak the comment to
/// A delegate that configures how a hierarchy of [DiagnosticsNode]s are
/// serialized by the Flutter Inspector.

File: packages/flutter/test/widgets/widget_inspector_test.dart [hunk 6]
```diff
@@ -2599,6 +2599,75 @@ class TestWidgetInspectorService extends Object with WidgetInspectorService {
...
+            (DiagnosticsNode node, InspectorSerializationDelegate delegate) {
+              final Map<String, Object> additionalJson = <String, Object>{};
+              final Object value = node.value;
+              if (value is Element) {
+                additionalJson['renderObject'] =
+                  value.renderObject.toDiagnosticsNode().toJsonMap(
+                    delegate.copyWith(subtreeDepth: 0),
+                  );
+              }
+              additionalJson['callbackIsCalled'] = true;
```
Comment (jacob314) [inline packages/flutter/test/widgets/widget_inspector_test.dart]: nit: callbackIsCalled --> 
callbackExecuted

File: packages/flutter/test/widgets/widget_inspector_test.dart [hunk 7]
```diff
@@ -2599,6 +2599,75 @@ class TestWidgetInspectorService extends Object with WidgetInspectorService {
...
+              return additionalJson;
+            },
+        );
+      final Map<String, Object> json = node.toJsonMap(delegate);
+      expect(json['callbackIsCalled'], true);
+      expect(json.containsKey('renderObject'), true);
+      expect(json['renderObject'], isA<Map<String, dynamic>>());
+      final Map<String, dynamic> renderObjectJson = json['renderObject'];
+      expect(renderObjectJson['description'], startsWith('RenderFlex'));
+      final InspectorSerializationDelegate unsafeDelegate =
```
Comment (jacob314) [inline packages/flutter/test/widgets/widget_inspector_test.dart]: nit: add a line break before lin 2652.

Comment (jacob314) [inline packages/flutter/test/widgets/widget_inspector_test.dart]: nit: unsafeDelegate to emptyDelegate.

**Review by jacob314 — COMMENTED** [review]
a few minor nits.
This functionality will be really useful for us to experiment with new Flutter Inspector features without landing code changes before the features are proven to be useful!

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 8]
```diff
@@ -2914,8 +2928,54 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
...
+  /// Callback to add additional experimental serialization properties.
+  ///
+  /// This callback can be used to customize the serialization of DiagnosticsNode
+  /// objects for experimental features in widget inspector clients such as
+  /// [Dart DevTools](https://github.com/flutter/devtools).
+  /// For example, [Dart DevTools](https://github.com/flutter/devtools)
+  /// can evaluate the following expression to register a VM Service API
+  /// with a custom serialization to experiment with visualizing layouts.
+  ///
+  /// The following code samples demonstrates adding the RenderObject associated
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: [RenderObject]

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 9]
```diff
@@ -2914,8 +2928,54 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
...
+  /// with an Element to the serialized data for all elements in the tree:
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: [Element]

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 10]
```diff
@@ -2914,8 +2928,54 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
...
+  /// ```dart
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: Add a blank line before this.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 11]
```diff
@@ -2889,20 +2889,34 @@ int _toLocationId(_Location location) {
...
     this.subtreeDepth = 1,
     this.includeProperties = false,
     @required this.service,
+    this.addAdditionalPropertiesCallback,
   });
 
+  /// Service used by GUI tools to interact with the [WidgetInspector].
   final WidgetInspectorService service;
+
+  /// The `groupName` parameter is not needed but is specified to regularize the
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: This doesn't really document what this property is for.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 12]
```diff
@@ -2914,8 +2928,54 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
...
+  ///       },
+  ///     ),
+  ///  );
+  /// }
+  /// ```
+  final Map<String, Object> Function(DiagnosticsNode, InspectorSerializationDelegate) addAdditionalPropertiesCallback;
+
   final List<DiagnosticsNode> _nodesCreatedByLocalProject = <DiagnosticsNode>[];
 
+  /// Whether the Inspector is interactive or not
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: Can you document what it means to be interactive?

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 13]
```diff
@@ -2947,7 +3010,7 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
     // we keep subtreeDepth from going down to zero until we reach nodes
     // that also exist in the summary tree. This ensures that every time
     // you expand a node in the details tree, you expand the entire subtree
-    // up until you reach the next nodes shared with the summary tree.
+    // up until you reach the next nodes shared with the sum mary tree.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: revert this?

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 14]
```diff
@@ -2914,8 +2928,54 @@ class _SerializationDelegate implements DiagnosticsSerializationDelegate {
...
+  /// with an Element to the serialized data for all elements in the tree:
+  /// ```dart
+  /// Map<String, Object> getDetailsSubtreeWithRenderObject(
+  ///   String id,
+  ///   String groupName,
+  ///   int subtreeDepth,
+  /// ) {
+  ///   return _nodeToJson(
+  ///     root,
+  ///     InspectorSerializationDelegate(
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: Where would I run this code? InspectorSerializationDelegate is marked as visibleForTesting, so would this only be in test code?

**Review by goderbauer — COMMENTED** [review]
Looks good. Just some doc nits.

File: packages/flutter/lib/src/widgets/widget_inspector.dart (hunk 12 above)
Comment (albertusangga) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: Will do!

File: packages/flutter/lib/src/widgets/widget_inspector.dart (hunk 11 above)
Comment (albertusangga) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: Alright, I agree.

File: packages/flutter/lib/src/widgets/widget_inspector.dart (hunk 13 above)
Comment (albertusangga) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: Thank you for catching this!

File: packages/flutter/lib/src/widgets/widget_inspector.dart (hunk 14 above)
Comment (jacob314) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: Debugging tools such as Dart DevTools can access it even if it was still private via the VM Service protocol. It is marked `@visibleForTesting` so we can test it from widget_inspector_test.dart

Comment (albertusangga) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: This code is run through the eval in the [Dart DevTools](https://github.com/flutter/devtools/pull/1396/files#diff-64bc585ceb9f869917465f175f0a9c6e). 
But currently in the flutter framework yes it is only in the test code.

**Comment by albertusangga** [comment]
@goderbauer ready for another round of review after improving the docs :-)

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**